download_ta_bdys.py

```diff
@@ -271,14 +255,192 @@ def main():
             name_field = ogr.FieldDefn('name', ogr.OFTString)
             name_field.SetWidth(100)
             output_lyr.CreateField(name_field)
-            
+
             pg_ds.ExecuteSQL("GRANT SELECT ON TABLE " + full_layer_name + " TO public")
         except Exception, e:
             logger.fatal('Can not create TA output table: %s' % (str(e)))
             sys.exit(1)
-    
+
+    #
+    # Retrieve all features from the ArcGIS endpoint
+    #
+
+    layer_output_srs = osr.SpatialReference()
+    layer_output_srs.ImportFromEPSG(layer_output_srid)
+    geojson_drv = ogr.GetDriverByName('GeoJSON')
+
+    if geojson_drv is None:
+        logger.fatal('Could not load the OGR GeoJSON driver')
+        sys.exit(1)
+
+    # ArcGIS server capabilities
+    capabilities_uri = base_uri + '/' + str(latest_layer) + '?f=json&pretty=true'
+
+    try:
+        response = urllib2.urlopen(capabilities_uri)
+    except Exception, e:
+        logger.fatal('Could open uri %s: %s' % (capabilities_uri, str(e)))
+        sys.exit(1)
+
+    try:
+        capabilities = json.load(response)
+    except Exception, e:
+        logger.fatal('Could load json from uri %s: %s' % (capabilities_uri, str(e)))
+        sys.exit(1)
+
+    # Get some of the capabilities
+    server_version = 0
+    server_pagination = False
+    if 'currentVersion' in capabilities.keys():
+        server_version = capabilities['currentVersion']
+    if 'advancedQueryCapabilities' in capabilities.keys():
+        if 'supportsPagination' in capabilities['advancedQueryCapabilities'].keys():
+            server_pagination = capabilities['advancedQueryCapabilities']['supportsPagination']
+    if 'displayField' in capabilities.keys():
+        display_field = capabilities['displayField']
+
+    # Feature count
+    count_uri = base_uri + '/' + str(latest_layer) + \
+             '/query?f=json&where=1=1&returnCountOnly=true'
+    try:
+        count_response = urllib2.urlopen(count_uri)
+    except Exception, e:
+        logger.fatal('Could open uri %s: %s' % (count_uri, str(e)))
+        sys.exit(1)
+
+    try:
+       count_json = json.load(count_response)
+       feat_count = int(count_json['count'])
+    except Exception, e:
+       logger.warning('Could not get feature count for uri %s: %s' % (count_uri, str(e)))
+
+    logger.info('Your OGR version is: ' + str(version_num))
+    logger.info('Stats NZ ArcGIS server version is : ' + str(server_version))
+    logger.info('Stats NZ ArcGIS server pagination flag is : ' + str(server_pagination))
+
+    # Temp layer to store features
+    tmp_drv = ogr.GetDriverByName("MEMORY")
+    tmp_ds = tmp_drv.CreateDataSource('memDs')
+    tmp_opn = tmp_drv.Open('memDs',1)
+    input_lyr = tmp_ds.CreateLayer('tmpLyr',srs=layer_output_srs,geom_type=ogr.wkbMultiPolygon)
+
+    # Use OGR to do paging if version 2.x and server supports paging 
+    if version_num >= 2000000 \
```

Comment:
Why under the GDAL 2.0 path is there manual code to do the paging? You should just use the built-in OGR paging. This is done by setting the GDAL config FEATURE_SERVER_PAGING option to YES (as you have done), setting the page size with the "resultRecordCount" HTTP KVP, and ensuring that the "resultOffset" KVP is not present in the URL